### PR TITLE
ostree: set rootflags when installing on btrfs (#1753485)

### PR DESCRIPTION
**Pull request by davide125**
When / is on a btrfs subvolume, we need to pass `rootflags` so it can be mounted correctly. This is an attempt at fixing https://bugzilla.redhat.com/show_bug.cgi?id=1753485

Full disclosure: I have no idea what I'm doing here, and this is completely untested so far.

### CODE REVIEW

File: pyanaconda/payload/rpmostreepayload.py [hunk 1]
```diff
@@ -450,6 +450,8 @@ def post_install(self):
             set_kargs_args = ["admin", "instutil", "set-kargs"]
             set_kargs_args.extend(bootloader.GetArguments())
             set_kargs_args.append("root=" + device_tree.GetFstabSpec(root_device))
+            if root_device.type == "btrfs subvolume":
+                set_kargs_args.append("rootflags=subvol=" + image.device.name)
```
Comment (mark-otaris) [inline pyanaconda/payload/rpmostreepayload.py:454]: Should possibly be:

```suggestion
                set_kargs_args.append("rootflags=subvol=" + root_device.name)
```

Comment (davide125) [reply pyanaconda/payload/rpmostreepayload.py:454]: Yes, definitely. Thanks!

**Comment by Conan-Kudo** [comment]
@davide125 You should probably add `Resolves: rhbz#1753485` as a second line in your commit message, per the Anaconda patch guidelines.

**Comment by poncovka** [comment]
Jenkins, test this please.

**Comment by jkonecny12** [comment]
Jenkins, it's ok to test.

**Comment by jkonecny12** [comment]
You can ignore the test result. Problem is a new version of pykickstart.

**Comment by felipeborges** [comment]
https://twitter.com/pid_eins/status/1281493584081027072
> That bug looks weird. You shouldn't have to set rootflags= to pick the default subvolume, that's just fragile. Use `btrfs subvol set-default` to set the default subvolume for a btrfs fs, so that this is implicit. I mean, it's great of rootflags= works, but requiring it sucks.

**Comment by Conan-Kudo** [comment]
Someone else yesterday brought up that our default partitioning arrangement "looks weird" too, and I need to find out why...

**Comment by poncovka** [comment]
@Conan-Kudo, can we merge this pull request or you want more time?

**Comment by davide125** [comment]
Can we merge this now to unblock silverblue? We can look at doing set-default as a followup (as that will require changing a bunch more places)

**Comment by Conan-Kudo** [comment]
@poncovka it's fine to merge now, some investigations will happen later and we can change as needed.

**Comment by Conan-Kudo** [comment]
> https://twitter.com/pid_eins/status/1281493584081027072
> 
> > That bug looks weird. You shouldn't have to set rootflags= to pick the default subvolume, that's just fragile. Use `btrfs subvol set-default` to set the default subvolume for a btrfs fs, so that this is implicit. I mean, it's great of rootflags= works, but requiring it sucks.

So I found out why we do this: upstream grub does not support reading the default subvolume. SUSE has patches for this in their GRUB2 package, some of those we have (hello from 2017 me!) and some of those we don't. So this is unfortunately the correct fix until we deal with that, and I don't know when we can get around to dealing with that...

At some point, I'd like to get the RH and SUSE bootloader folks to come together on maybe getting us to BLS for btrfs boot to snapshot (which is the feature enabled by these patches that would help in this case), because this Jenga tower of patches is bonkers...

**Comment by martinezjavier** [comment]
> So I found out why we do this: upstream grub does not support reading the default subvolume. SUSE has patches for this in their GRUB2 package, some of those we have (hello from 2017 me!) and some of those we don't. So this is unfortunately the correct fix until we deal with that, and I don't know when we can get around to dealing with that...

If you point me to those patches, I can include them in our package if that helps. Or you could propose a pull request for the https://github.com/rhboot/grub2/tree/fedora-33 branch.

> At some point, I'd like to get the RH and SUSE bootloader folks to come together on maybe getting us to BLS for btrfs boot to snapshot (which is the feature enabled by these patches that would help in this case), because this Jenga tower of patches is bonkers...

Nobody is happy with having so many patches on top of mainline. Specially since most distros are carrying more or less the same set. 

But things have been improving lately and upstream got much better at reviewing the posted patches. We managed to get ~40 patches merged for the upcoming 2.06 release. So we will get there, it just takes time.

**Comment by poncovka** [comment]
Hi, I have updated the pull request, because the installation was failing. The current code seems to work fine.

**Comment by Conan-Kudo** [comment]
> > So I found out why we do this: upstream grub does not support reading the default subvolume. SUSE has patches for this in their GRUB2 package, some of those we have (hello from 2017 me!) and some of those we don't. So this is unfortunately the correct fix until we deal with that, and I don't know when we can get around to dealing with that...
> 
> If you point me to those patches, I can include them in our package if that helps. Or you could propose a pull request for the https://github.com/rhboot/grub2/tree/fedora-33 branch.
> 
> > At some point, I'd like to get the RH and SUSE bootloader folks to come together on maybe getting us to BLS for btrfs boot to snapshot (which is the feature enabled by these patches that would help in this case), because this Jenga tower of patches is bonkers...
> 
> Nobody is happy with having so many patches on top of mainline. Specially since most distros are carrying more or less the same set.
> 
> But things have been improving lately and upstream got much better at reviewing the posted patches. We managed to get ~40 patches merged for the upcoming 2.06 release. So we will get there, it just takes time.

@martinezjavier The patches can be found here: https://build.opensuse.org/package/show/Base:System/grub2

You can see _only_ the Btrfs patches by putting "btrfs" into the Search filter on the file list.

**Comment by Conan-Kudo** [comment]
> > So I found out why we do this: upstream grub does not support reading the default subvolume. SUSE has patches for this in their GRUB2 package, some of those we have (hello from 2017 me!) and some of those we don't. So this is unfortunately the correct fix until we deal with that, and I don't know when we can get around to dealing with that...
> 
> If you point me to those patches, I can include them in our package if that helps. Or you could propose a pull request for the https://github.com/rhboot/grub2/tree/fedora-33 branch.
> 
> > At some point, I'd like to get the RH and SUSE bootloader folks to come together on maybe getting us to BLS for btrfs boot to snapshot (which is the feature enabled by these patches that would help in this case), because this Jenga tower of patches is bonkers...
> 
> Nobody is happy with having so many patches on top of mainline. Specially since most distros are carrying more or less the same set.
> 
> But things have been improving lately and upstream got much better at reviewing the posted patches. We managed to get ~40 patches merged for the upcoming 2.06 release. So we will get there, it just takes time.

@martinezjavier The patches can be found here: https://build.opensuse.org/package/show/Base:System/grub2

You can see _only_ the Btrfs patches by putting "btrfs" into the Search filter on the file list. The patches were made by @jeffmahoney and @WenhuaChang primarily.

**Comment by davide125** [comment]
Thanks @poncovka, the updated PR lgtm.

**Pull request merged**